README.md
# Flexible Polyline Encoding for R <img src="man/figures/logo.png" align="right" alt="" width="120" />

<!-- badges: start -->

[![CRAN status](https://www.r-pkg.org/badges/version/flexpolyline)](https://CRAN.R-project.org/package=flexpolyline)
[![CRAN checks](https://badges.cranchecks.info/worst/flexpolyline.svg)](https://cran.r-project.org/web/checks/check_results_flexpolyline.html)
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/last-month/flexpolyline?color=brightgreen)](https://CRAN.R-project.org/package=flexpolyline)
[![R-CMD-check](https://github.com/munterfi/flexpolyline/actions/workflows/check-standard.yaml/badge.svg)](https://github.com/munterfi/flexpolyline/actions/workflows/check-standard.yaml)
[![Codecov test coverage](https://codecov.io/gh/munterfi/flexpolyline/branch/master/graph/badge.svg)](https://app.codecov.io/gh/munterfi/flexpolyline?branch=master)
[![CodeFactor](https://www.codefactor.io/repository/github/munterfi/flexpolyline/badge)](https://www.codefactor.io/repository/github/munterfi/flexpolyline)

<!-- badges: end -->

The **flexpolyline** R package provides a binding to the
[C++ implementation](https://github.com/heremaps/flexible-polyline/tree/master/cpp) of the
flexible polyline encoding by [HERE](https://github.com/heremaps/flexible-polyline).
The flexible polyline encoding is a lossy compressed representation of a list of
coordinate pairs or coordinate triples. The encoding is achieved by:
(1) Reducing the decimal digits of each value;
(2) encoding only the offset from the previous point;
(3) using variable length for each coordinate delta; and
(4) using 64 URL-safe characters to display the result.
The flexible polyline encoding is a variant of the [Encoded Polyline Algorithm Format](https://developers.google.com/maps/documentation/utilities/polylinealgorithm) by Google.

**Note:**

- The order of the coordinates (lng, lat) does not correspond to the original
  C++ implementation (lat, lng). This enables direct conversion to **sf**
  objects, without reordering the columns.
- The encoding is lossy, this means the encoding process could reduce the
  precision of your data.

## Installation

Install the released version of **flexpolyline** from [CRAN](https://CRAN.R-project.org/package=flexpolyline):

```r
install.packages("flexpolyline")
```

To install the development version from [GitHub](https://github.com/munterfi/flexpolyline):

```r
remotes::install_github("munterfi/flexpolyline")
```

## C++ binding

Encoding and decoding in R is straight forward by using `encode()` and
`decode()`. These functions are binding to the flexpolyline C++ implementation
and reflect the arguments and return values of their counterparts
(`hf::flexpolyline::polyline_encode` and `hf::flexpolyline::polyline_decode`):

```r
line <- matrix(
  c(8.69821, 50.10228, 10,
    8.69567, 50.10201, 20,
    8.69150, 50.10063, 30,
    8.68752, 50.09878, 40),
  ncol = 3, byrow = TRUE
)

encode(line)

decode("BlBoz5xJ67i1BU1B7PUzIhaUxL7YU")
```

## Simple feature support

The **flexpolyline** package integrates with the popular
[sf](https://CRAN.R-project.org/package=sf) package for working with spatial
data in R, which is built on the concept of simple features. The package
provides additional functions for encoding and decoding sf objects directly:

```r
sfg <- sf::st_linestring(line, dim = "XYZ")

encode_sf(sfg)

decode_sf("BlBoz5xJ67i1BU1B7PUzIhaUxL7YU")
```

## Contributing

- Contributions to this package are welcome, issues and pull requests are the
  preferred ways to share them. Please see the [Contribution Guidelines](https://github.com/munterfi/flexpolyline/blob/master/.github/CONTRIBUTING.md).
- This project is released with a [Contributor Code of Conduct](https://github.com/munterfi/flexpolyline/blob/master/.github/CODE_OF_CONDUCT.md).
  By participating in this project you agree to abide by its terms.

## References

- [Flexible Polyline Encoding by HERE](https://github.com/heremaps/flexible-polyline)
- [Encoded Polyline Algorithm Format](https://developers.google.com/maps/documentation/utilities/polylinealgorithm)
- [Simple Features for R](https://CRAN.R-project.org/package=sf)
- Inspired by the [googlePolylines](https://github.com/SymbolixAU/googlePolylines) package

## License

The **flexpolyline** R package is licensed under GNU GPL v3.0. The C++
implementation by HERE Europe B.V. is licensed under MIT.


NEWS.md
# flexpolyline 1.0.0

- Upgrade to current version of the flexpolyline [C++ implementation] (https://github.com/heremaps/flexible-polyline/blob/3e1ebc2e2ce39111b0e2e91260b98053ac965f77/cpp/include/hf/flexpolyline.h).
- Introduce encoder and decoder classes in C++ to separate responsibilities and improve error messages.
- Remove function set (`set_third_dimension()`) the third dimension type of a flexible polyline encoded string, as it can be done using `decode()` and `encode()`.
- Split tests into encoding and decoding and add test helpers.
- Upgraded **testthat** edition by calling `usethis::use_testthat(3)`.
- Replaced github actions with current versions from **r-lib/actions**.

# flexpolyline 0.3.0

- Update C++11 to current CRAN default of C++17.

# flexpolyline 0.2.7

- Fix missing missing inclusion of `<cstdint>`, closes [#58](https://github.com/munterfi/flexpolyline/issues/58).
- Resubmit to CRAN due to since the package was archived on 2022-12-27 as email to the maintainer was temporarily not deliverable.
- Updated r-lib/actions in github action `R-CMD-check.yaml` and `pkgdown.yaml`.

# flexpolyline 0.2.6

- Switch to HTML5 for documentation pages, closes [#56](https://github.com/munterfi/flexpolyline/issues/56).

# flexpolyline 0.2.5

- Changed workflow from git flow to trunc-based development. Removed develop branch. New features are merged directly into `master` and releases are tagged after CRAN has accepted the package submission, which triggers the documentation build.

# flexpolyline 0.2.4

- Adjust URLs to GitHub account due to renaming munterfinger to @munterfi.

# flexpolyline 0.2.3

- Skip encoding and decoding validation tests on CRAN due to minor deviations in the results due to rounding (half up vs. half even) on different platforms (e.g. Apple M1) at high precision values (closes [#49](https://github.com/munterfi/flexpolyline/issues/49))

# flexpolyline 0.2.2

- Fix wrong integer shift resulting in lost bits for precision values greater than 7 (see [heremaps/flexible-polyline#36](https://github.com/heremaps/flexible-polyline/issues/36), closes [#44](https://github.com/munterfi/flexpolyline/issues/44)).
- Fix CRAN note for a specified lazy data statement without data directory (closes [#43](https://github.com/munterfi/flexpolyline/issues/43)).
- Use **styler** package and use `tyler::tidyverse_style()`to format the package.
- Solve **lintr** issues, except from line length issues (limit of 80 characters).

# flexpolyline 0.2.1

- Skip input tests of C++ binding on CRAN, closes #35.
- Added contribution guidelines, code of conduct and issue templates.

# flexpolyline 0.2.0

- Fix clang range-loop-analysis warning on macOS in `flexpolyline.h` (Apple clang version 12.0.0).
- Support for geometry types `"POLYGON"` and `"POINT"` in `encode_sf()` and `decode_sf()`, closes #31.
- Added functions to get (`get_third_dimension()`) and set (`set_third_dimension()`) the third dimension type of a flexible polyline encoded string.
- Sign in to CodeFactor.io and add badge to continuously track code quality.
- Use exception classes when throwing an exception in C++.
- Improve coverage of tests.

# flexpolyline 0.1.1

- Add ORCID to author field in `DESCRIPTION`.
- Limit the encoding check in the C++ binding test to 7 digits.
- Use explicit type casts before left shifting and reassigning (`x <<= y`) to avoid UBSAN runtime error _'left shift of negative value'_ in `flexpolyline.h`.

# flexpolyline 0.1.0

First release of the **flexpolyline** package, which provides a binding to the
[C++ implementation](https://github.com/heremaps/flexible-polyline/tree/master/cpp) of the
flexible polyline encoding by [HERE](https://github.com/heremaps/flexible-polyline).
The flexible polyline encoding is a lossy compressed representation of a list of
coordinate pairs or coordinate triples. The encoding is achieved by:

(1) Reducing the decimal digits of each value;
(2) encoding only the offset from the previous point;
(3) using variable length for each coordinate delta; and
(4) using 64 URL-safe characters to display the result.

The felxible polyline encoding is a variant of the [Encoded Polyline Algorithm Format](https://developers.google.com/maps/documentation/utilities/polylinealgorithm) by Google.

**License:**

- The **flexpolyline** R package is licensed under GNU GPL v3.0.
- The C++ implementation by HERE Europe B.V. is licensed under MIT.


R/encode_sf.R
#' Wrapper function for encoding simple features
#'
#' A wrapper function for \code{\link{encode}} that converts simple feature geometries
#' of the sf package to flexible polyline encoded strings.
#'
#' @param geom simple feature, \code{sf}, \code{sfc} or \code{sfg} object with geometry type \code{"POINT"}, \code{"LINESTRING"} or \code{"POLYGON"}.
#' @param precision integer, precision to use in encoding (between 0 and 15, \code{default=5}).
#' @param third_dim integer, type of the third dimension (0: ABSENT, 1: LEVEL, 2: ALTITUDE, 3: ELEVATION, 4, 6: CUSTOM1, 7: CUSTOM2, \code{default=NULL}).
#' @param third_dim_precision integer, precision to use in encoding for the third dimension (between 1 and 15, \code{default=precision}).
#'
#' @return
#' The line as string in the flexible polyline encoding format.
#'
#' @export
#'
#' @examples
#' # 3D point
#' point3d <- sf::st_point(
#'   matrix(c(8.69821, 50.10228, 10), ncol = 3, byrow = TRUE),
#'   dim = "XYZ"
#' )
#' encode_sf(point3d)
#'
#' # 2D linestring
#' line2d <- sf::st_linestring(
#'   matrix(c(
#'     8.69821, 50.10228,
#'     8.69567, 50.10201,
#'     8.68752, 50.09878
#'   ), ncol = 2, byrow = TRUE)
#' )
#' encode_sf(line2d)
#'
#' # 3D polygon
#' poly3d <- sf::st_polygon(list(
#'   matrix(c(
#'     8.69821, 50.10228, 10,
#'     8.69567, 50.10201, 20,
#'     8.69150, 50.10063, 30,
#'     8.69821, 50.10228, 10
#'   ), ncol = 3, byrow = TRUE)
#' ), dim = "XYM")
#' encode_sf(poly3d)
encode_sf <- function(geom, precision = 5, third_dim = NULL,
                      third_dim_precision = precision) {
  UseMethod("encode_sf", geom)
}

#' @export
encode_sf.sfg <- function(geom, precision = 5, third_dim = NULL,
                          third_dim_precision = precision) {
  if (!sf::st_geometry_type(geom) %in% c("POINT", "LINESTRING", "POLYGON")) {
    stop(
      sprintf(
        "Invalid geometry type '%s' of input, only 'POINT', 'LINESTRING' and 'POLYGON' is supported.",
        sf::st_geometry_type(geom)
      )
    )
  }
  if (class(geom)[1] == "XY") {
    third_dim <- 0
    encoded <- encode(
      sf::st_coordinates(geom)[, c(1:2), drop = FALSE],
      precision, third_dim, third_dim_precision
    )
  } else {
    if (is.null(third_dim)) {
      if (class(geom)[1] == "XYZ") {
        third_dim <- 3
      } else {
        third_dim <- 6
      }
    }
    encoded <- encode(
      sf::st_coordinates(geom)[, c(1:3), drop = FALSE],
      precision, third_dim, third_dim_precision
    )
  }
  return(encoded)
}

#' @export
encode_sf.sfc <- function(geom, precision = 5, third_dim = NULL,
                          third_dim_precision = precision) {
  return(
    sapply(geom, function(x) {
      encode_sf.sfg(x, precision, third_dim, third_dim_precision)
    })
  )
}

#' @export
encode_sf.sf <- function(geom, precision = 5, third_dim = NULL,
                         third_dim_precision = precision) {
  return(
    encode_sf.sfc(
      sf::st_geometry(geom),
      precision, third_dim, third_dim_precision
    )
  )
}
